primitives/src/currency.rs

```diff
@@ -242,26 +289,54 @@ impl TryFrom<[u8; 32]> for CurrencyId {
 	}
 }
 
-impl TryFrom<CurrencyId> for u32 {
-	type Error = ();
-
-	fn try_from(val: CurrencyId) -> Result<Self, Self::Error> {
+impl From<DexShare> for u32 {
+	fn from(val: DexShare) -> u32 {
 		let mut bytes = [0u8; 4];
 		match val {
-			CurrencyId::Token(token) => {
+			DexShare::Token(token) => {
 				bytes[3] = token.into();
 			}
-			CurrencyId::Erc20(address) => {
+			DexShare::Erc20(address) => {
 				let is_zero = |&&d: &&u8| -> bool { d == 0 };
 				let leading_zeros = address.as_bytes().iter().take_while(is_zero).count();
 				let index = if leading_zeros > 16 { 16 } else { leading_zeros };
 				bytes[..].copy_from_slice(&address[index..index + 4][..]);
 			}
-			_ => {
-				return Err(());
-			}
 		}
-		Ok(u32::from_be_bytes(bytes))
+		u32::from_be_bytes(bytes)
+	}
+}
+
+/// Generate the EvmAddress from CurrencyId so that evm contracts can call the erc20 contract.
+impl TryFrom<CurrencyId> for EvmAddress {
+	type Error = ();
+
+	fn try_from(val: CurrencyId) -> Result<Self, Self::Error> {
+		let address = match val {
```

Comment:
```suggestion
		let maybe_address = match val {
```

We should try to avoid any shadowed variables.